src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,1339 @@
+  }
+};
+
+static inline std::ostream& operator<<(std::ostream& os, const fixed_key_3& k) {
+  if (k.snap != CEPH_NOSNAP) {
+    os << "s" << k.snap << ",";
+  }
+  if (k.gen != ghobject_t::NO_GEN) {
+    os << "g" << k.gen << ",";
+  }
+  return os;
+}
+
+// all elements in this node must share the same <shard, pool, hash, namespace, oid>
+// but the unlike other FixedKey<>, a node with FixedKey<3> does not have
+// variable_sized_key, so if it is an inner node, we can just embed the child
+// addr right in the key.
+template<>
+struct FixedKey<3, ntype_t::inner> : public fixed_key_3 {
+  static constexpr int level = 3;
+  // the item is embedded in the key
+  laddr_t child_addr = 0;
+
+  FixedKey() = default;
+  void set(const ghobject_t& oid, laddr_t new_child_addr) {
+    update_with_oid(oid);
+    child_addr = new_child_addr;
+  }
+  // unlikely get called, though..
+  void update(const ghobject_t& oid) {}
+  template<int N>
+  std::enable_if_t<N < 3> set(const FixedKey<N, ntype_t::inner>&,
+                              laddr_t new_child_addr) {
+    child_addr = new_child_addr;
+  }
+  void set(const FixedKey& k, laddr_t new_child_addr) {
+    snap = k.snap;
+    gen = k.gen;
+    child_addr = new_child_addr;
+  }
+  void set(const variable_sized_key_t& k, laddr_t new_child_addr) {
+    snap = k.snap;
+    gen = k.gen;
+    child_addr = new_child_addr;
+  }
+};
+
+template<>
+struct FixedKey<3, ntype_t::leaf> : public fixed_key_3 {
+  static constexpr int level = 3;
+  uint16_t offset = 0;
+
+  FixedKey() = default;
+  void set(const ghobject_t& oid, uint16_t new_offset) {
+    update_with_oid(oid);
+    offset = new_offset;
+  }
+  void set(const FixedKey& k, uint16_t new_offset) {
+    snap = k.snap;
+    gen = k.gen;
+    offset = new_offset;
+  }
+  template<int N>
+  std::enable_if_t<N < 3> set(const FixedKey<N, ntype_t::leaf>&,
+                              uint16_t new_offset) {
+    offset = new_offset;
+  }
+};
+
+// prefix-compress the key
+struct prefix_compressed_key_t {
+  struct index_t {
+    enum {
+      shared_key_len,
+      unshared_key_len,
+      onode_len,
+      key_data,
+      onode_data,
+    };
+  };
+  using layout_type = asci::Layout<uint8_t,
+                                   uint8_t,
+                                   uint16_t,
+                                   char,
+                                   char>;
+  constexpr static layout_type cell_layout(uint8_t unshared_len,
+                                           uint16_t onode_len) {
+    return layout_type{1, 1, 1, unshared_len, onode_len};
+  }
+};
+
+struct tag_t {
+  template<int N, ntype_t node_type>
+  static constexpr tag_t create() {
+    static_assert(std::clamp(N, 0, 3) == N);
+    return tag_t{N, static_cast<uint8_t>(node_type)};
+  }
+  bool is_leaf() const {
+    return ntype_t{node_type} == ntype_t::leaf;
+  }
+  int layout() const {
+    return layout_type;
+  }
+  int layout_type : 4;
+  uint8_t node_type : 4;
```

Comment:
the layout_type is defined as `int` because i want to be consistent. in https://github.com/tchaikov/ceph/commit/9cd11f210d7816a3f36ad703c7454a9f337d40d1#diff-31b6a4983be200b0524f43d2735cbc1bR100. i am comparing `layout_type` with 0, as i am decrementing it. and i define the parent node type like
```c++
  // (-1, N]
  using parent_node_variant_t = typename parent_node_variant<-1, N>::type;
```

in hope to simplify the definition of implementation. we can revise it if it's not intuitive or not correct.

the reason why i define `node_type` as `uint8_t` is that it's just a flag, and the sign does not matter.